[GStreamer] New test media/media-source/media-source-webm-configuration-change.html failing

philn opened the pull request:
#### 7734fe565b9151fb97688649bd2a53c14327e55a
<pre>
[GStreamer] New test media/media-source/media-source-webm-configuration-change.html failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=253832">https://bugs.webkit.org/show_bug.cgi?id=253832</a>

Reviewed by Xabier Rodriguez-Calvar.

The append pipeline is now parsing VP9, which allows the corresponding video track to advertise a
more accurate codec string. The BT2020 matrix coefficients mapping were also updated to the
Conventional non-constant luminance Y&apos;C&apos;BC&apos;R defined in Rec. ITU-R BT.2020-2, I opened an issue in
GStreamer&apos;s gitlab regarding this, <a href="https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/3156.">https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/3156.</a>

* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/platform/graphics/gstreamer/AudioTrackPrivateGStreamer.cpp:
(WebCore::AudioTrackPrivateGStreamer::updateConfigurationFromCaps):
* Source/WebCore/platform/graphics/gstreamer/AudioTrackPrivateGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::videoColorSpaceFromInfo):
(WebCore::fillVideoInfoColorimetryFromColorSpace):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::updateTracks):
* Source/WebCore/platform/graphics/gstreamer/TrackPrivateBaseGStreamer.h:
(WebCore::TrackPrivateBaseGStreamer::updateConfigurationFromCaps):
* Source/WebCore/platform/graphics/gstreamer/VideoTrackPrivateGStreamer.cpp:
(WebCore::VideoTrackPrivateGStreamer::updateConfigurationFromCaps):
* Source/WebCore/platform/graphics/gstreamer/VideoTrackPrivateGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::createOptionalParserForFormat):

Canonical link: <a href="https://commits.webkit.org/271224@main">https://commits.webkit.org/271224@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8bec8a02437a9323d41d501e400fc52cc5aa2688

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27681 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6319 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28927 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29901 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25302 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28156 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8260 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3712 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/25061 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27947 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5077 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23750 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4404 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4576 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24757 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30541 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25278 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25185 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30693 "Found 1 new test failure: fast/events/monotonic-event-time.html (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4595 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2742 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28663 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6092 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6658 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5048 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5023 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/AudioTrackPrivateGStreamer.h:
```diff
@@ -60,9 +60,10 @@ class AudioTrackPrivateGStreamer final : public AudioTrackPrivate, public TrackP
     AtomString label() const final { return m_label; }
     AtomString language() const final { return m_language; }
 
+    void updateConfigurationFromCaps(const GRefPtr<GstCaps>&&) final;
```
Wow, I guess you just moved the function, but const && usually makes no sense. If I reviewed this in the past I apologize for missing this. You either pass
* const &:  to state that only const operations or a copy should be made
* non const &&: to state that what you are passing is the full ownership of the object.

Please change this, everywhere. No const &&, please. If it's too much for this patch, please do it before in a separate patch.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/AudioTrackPrivateGStreamer.h:
```diff
@@ -60,9 +60,10 @@ class AudioTrackPrivateGStreamer final : public AudioTrackPrivate, public TrackP
 protected:
-    void updateConfigurationFromCaps(const GRefPtr<GstCaps>&&) override;
-    void updateConfigurationFromTags(const GRefPtr<GstTagList>&&) override;
+    void updateConfigurationFromTags(const GRefPtr<GstTagList>&&) final;
```
ditto.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1501,14 +1501,15 @@ void MediaPlayerPrivateGStreamer::updateTracks(const GRefPtr<GstObject>& collect
             m_##type##Tracks.add(streamId, WTFMove(track));             \
         }                                                               \
         auto track = m_##type##Tracks.get(streamId);                    \
+        track->updateConfigurationFromCaps(WTFMove(caps));              \
```
This makes me a bit uneasy because it's a macro but in the end it resembles a function. I think you should move in the caller and avoid moving here.

philn replied:
I don't think that's possible, the macro is not a function.

philn replied:
Doesn't need to be a macro argument actually.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/TrackPrivateBaseGStreamer.h:
```diff
@@ -69,6 +69,8 @@ class TrackPrivateBaseGStreamer {
 
     static String trackIdFromPadStreamStartOrUniqueID(TrackType, unsigned index, const GRefPtr<GstPad>&);
 
+    virtual void updateConfigurationFromCaps(const GRefPtr<GstCaps>&&) { }
```
ditto

calvaris commented on Source/WebCore/platform/graphics/gstreamer/TrackPrivateBaseGStreamer.h:
```diff
@@ -81,7 +83,6 @@ class TrackPrivateBaseGStreamer {
     virtual void capsChanged(const String&, GRefPtr<GstCaps>&&) { }
     void installUpdateConfigurationHandlers();
-    virtual void updateConfigurationFromCaps(const GRefPtr<GstCaps>&&) { }
     virtual void updateConfigurationFromTags(const GRefPtr<GstTagList>&&) { }
```
ditto.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/VideoTrackPrivateGStreamer.h:
```diff
@@ -61,9 +61,10 @@ class VideoTrackPrivateGStreamer final : public VideoTrackPrivate, public TrackP
     AtomString label() const final { return m_label; }
     AtomString language() const final { return m_language; }
 
+    void updateConfigurationFromCaps(const GRefPtr<GstCaps>&&) final;
```
ditto.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/VideoTrackPrivateGStreamer.h:
```diff
@@ -61,9 +61,10 @@ class VideoTrackPrivateGStreamer final : public VideoTrackPrivate, public TrackP
 protected:
-    void updateConfigurationFromCaps(const GRefPtr<GstCaps>&&) override;
-    void updateConfigurationFromTags(const GRefPtr<GstTagList>&&) override;
+    void updateConfigurationFromTags(const GRefPtr<GstTagList>&&) final;
```
ditto.

philn committed:
[GStreamer] New test media/media-source/media-source-webm-configuration-change.html failing
https://bugs.webkit.org/show_bug.cgi?id=253832

Reviewed by Xabier Rodriguez-Calvar.

The append pipeline is now parsing VP9, which allows the corresponding video track to advertise a
more accurate codec string. The BT2020 matrix coefficients mapping were also updated to the
Conventional non-constant luminance Y'C'BC'R defined in Rec. ITU-R BT.2020-2, I opened an issue in
GStreamer's gitlab regarding this, https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/3156.

* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/platform/graphics/gstreamer/AudioTrackPrivateGStreamer.cpp:
(WebCore::AudioTrackPrivateGStreamer::updateConfigurationFromCaps):
* Source/WebCore/platform/graphics/gstreamer/AudioTrackPrivateGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::videoColorSpaceFromInfo):
(WebCore::fillVideoInfoColorimetryFromColorSpace):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::updateTracks):
* Source/WebCore/platform/graphics/gstreamer/TrackPrivateBaseGStreamer.h:
(WebCore::TrackPrivateBaseGStreamer::updateConfigurationFromCaps):
* Source/WebCore/platform/graphics/gstreamer/VideoTrackPrivateGStreamer.cpp:
(WebCore::VideoTrackPrivateGStreamer::updateConfigurationFromCaps):
* Source/WebCore/platform/graphics/gstreamer/VideoTrackPrivateGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::createOptionalParserForFormat):

Canonical link: https://commits.webkit.org/271224@main
